crates/types/src/predicate/header.rs

```diff
@@ -0,0 +1,598 @@
+
+        // Check the decoded header is consistent with the fixed size part.
+        header.check_consistency(&fh)?;
+
+        // Check the bounds of the predicate.
+        let bounds = PredicateBounds {
+            num_state_reads,
+            num_constraints,
+            state_read_lens: header.state_reads.iter().map(ExactSizeIterator::len),
+            constraint_lens: header.constraints.iter().map(ExactSizeIterator::len),
+        };
+        check_predicate_bounds(bounds)?;
+
+        Ok(header)
+    }
+
+    /// The number of bytes this header points to.
+    ///
+    /// This includes the [`EncodedFixedSizeHeader::SIZE`]
+    /// and the lengths of the programs as well as the programs themselves.
+    pub fn bytes_len(&self) -> usize {
+        let sr = self
+            .state_reads
+            .iter()
+            .fold(0usize, |acc, p| acc.saturating_add(p.len()));
+
+        let c = self
+            .constraints
+            .iter()
+            .fold(0usize, |acc, p| acc.saturating_add(p.len()));
+
+        // Two bytes per length.
+        let lens = self
+            .state_reads
+            .len()
+            .saturating_add(self.constraints.len())
+            .saturating_mul(core::mem::size_of::<u16>());
+
+        EncodedFixedSizeHeader::SIZE
+            .saturating_add(sr)
+            .saturating_add(c)
+            .saturating_add(lens)
+    }
+
+    /// Decode the state read programs from a buffer.
+    /// This simply re-nests the programs from the flat buffer.
+    /// The underlying data is not modified.
+    ///
+    /// # Panics
+    /// Panics if the buffer is too small.
+    /// Lengths must be checked before calling this method.
+    /// Check with [`Self::bytes_len`].
+    pub fn decode_state_read(&self, buf: &[u8]) -> Vec<Vec<u8>> {
+        self.state_reads
+            .iter()
+            .map(|range| buf[range.clone()].to_vec())
+            .collect()
+    }
+
+    /// Decode the constraint check programs from a buffer.
+    /// This simply re-nests the programs from the flat buffer.
+    /// The underlying data is not modified.
+    ///
+    /// # Panics
+    /// Panics if the buffer is too small.
+    /// Lengths must be checked before calling this method.
+    /// Check with [`Self::bytes_len`].
+    pub fn decode_constraints(&self, buf: &[u8]) -> Vec<Vec<u8>> {
+        self.constraints
+            .iter()
+            .map(|range| buf[range.clone()].to_vec())
+            .collect()
+    }
+
+    /// Check the [`DecodedHeader`] is consistent with the [`FixedSizeHeader`].
+    pub fn check_consistency(&self, header: &FixedSizeHeader) -> DecodeResult<()> {
+        if self.state_reads.len() != header.num_state_reads as usize {
+            return Err(DecodeError::IncorrectBodyLength);
+        }
+        if self.constraints.len() != header.num_constraints as usize {
+            return Err(DecodeError::IncorrectBodyLength);
+        }
+        Ok(())
+    }
+}
```

Comment:
Same for this impl w.r.t. visibility - are we sure we want/need to expose everything? :thinking: Mostly just thinking that not exposing may give us some flexibility to optimise a bit more in the future (e.g with zero-allocation approaches)